pkg/storage/couchdb/couchdbstore.go

```diff
@@ -0,0 +1,193 @@
+
+const (
+	blankHostErrMsg           = "hostURL for new CouchDB provider can't be blank"
+	failToCloseProviderErrMsg = "failed to close provider"
+)
+
+// NewProvider instantiates Provider
+func NewProvider(hostURL string) (*Provider, error) {
+	if hostURL == "" {
+		return nil, errors.New(blankHostErrMsg)
+	}
+
+	client, err := kivik.New("couch", hostURL)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Provider{hostURL: hostURL, couchDBClient: client, dbs: map[string]*CouchDBStore{}}, nil
+}
+
+// CreateStore creates a new store with the given name.
+func (p *Provider) CreateStore(name string) error {
+	err := p.couchDBClient.CreateDB(context.Background(), name)
+	return err
+}
+
+// OpenStore opens an existing store with the given name and returns it.
+func (p *Provider) OpenStore(name string) (storage.Store, error) {
+	exists, err := p.storeExists(name)
+	if err != nil {
+		return nil, err
+	}
+
+	if !exists {
+		return nil, storage.ErrStoreNotFound
+	}
+
+	db := p.couchDBClient.DB(context.Background(), name)
+
+	if dbErr := db.Err(); dbErr != nil {
+		return nil, dbErr
+	}
+
+	store := &CouchDBStore{db: db}
+
+	p.dbs[name] = store
+
+	return store, nil
+}
+
+func (p *Provider) storeExists(name string) (bool, error) {
+	_, existsInCache := p.dbs[name]
+
+	if existsInCache {
+		return true, nil
+	}
+
+	return p.couchDBClient.DBExists(context.Background(), name)
+}
+
+// CloseStore closes a previously opened store.
+func (p *Provider) CloseStore(name string) error {
+	store, exists := p.dbs[name]
+	if !exists {
+		return storage.ErrStoreNotFound
+	}
+
+	delete(p.dbs, name)
+
+	return store.db.Close(context.Background())
+}
+
+// Close closes the provider.
+func (p *Provider) Close() error {
+	for _, store := range p.dbs {
+		err := store.db.Close(context.Background())
+		if err != nil {
+			return fmt.Errorf(failToCloseProviderErrMsg+": %w", err)
+		}
+	}
+
+	p.dbs = make(map[string]*CouchDBStore)
+
+	return p.couchDBClient.Close(context.Background())
+}
+
+// CouchDBStore represents a CouchDB-backed database.
+type CouchDBStore struct {
+	db *kivik.DB
+}
+
+// Put stores the given key-value pair in the store.
+func (c *CouchDBStore) Put(k string, v []byte) error {
+	var valueToPut []byte
+	if isJSON(v) {
+		valueToPut = v
+	} else {
+		valueToPut = wrapTextAsCouchDBAttachment(v)
+	}
+
+	_, err := c.db.Put(context.Background(), k, valueToPut)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func isJSON(textToCheck []byte) bool {
+	var js json.RawMessage
+	return json.Unmarshal(textToCheck, &js) == nil
+}
+
+// Kivik has a PutAttachment method, but it requires creating a document first and then adding an attachment after.
+// We want to do it all in one step, hence this manual stuff below.
+func wrapTextAsCouchDBAttachment(textToWrap []byte) []byte {
+	encodedTextToWrap := base64.StdEncoding.EncodeToString(textToWrap)
+	return []byte(`{"_attachments": {"data": {"data": "` + encodedTextToWrap + `", "content_type": "text/plain"}}}`)
```

Comment:
will these need indexing?

Reply:
No, they shouldn't. This whole attachment thing is just to allow the CouchDB store to work with non-JSON payloads. The primary use case for the CouchDB store is to store EDV EncryptedDocuments, which are JSON. Those are the ones that will need indexing, and they won't be stored as attachments, so we should be okay.